### PR TITLE
`arrow-reflect`. Run-time introspection for Arrow patterns + Ank and Docs improvements

### PR DESCRIPTION
**Pull request by raulraja**
Includes:

- [x] `arrow-reflect`, a new module in charge of providing meta-programing facilities related to the introspection of Arrow patterns such as Type class instances or higher-kinded types.
- [x] `kotlin:ank:replace`. A new Ank command that allows replacing an entire snippet with the output of the resulting evaluation.

Combining `arrow-reflect` and `kotlin:ank:replace` we can now automate in the Arrow documentation the list of type classes and data types a given type class or data type supports:

### Supported type classes
```
` ` `kotlin:ank:replace
import arrow.reflect.*
import arrow.data.*
import arrow.core.*

DataType(Either::class).tcMarkdownList()
` ` `
```

generates a markdown table like the one below in place of the code snippet:

```
### Supported type classes

| Module | Type classes |
| --- | --- |
|__arrow.aql__|[Count](link), [From](link), [GroupBy](link), [OrderBy](link), [Select](link), [Sum](link), [Union](link)|
|__arrow.optics.typeclasses__|[Each](link)|
|__arrow.typeclasses__|[ApplicativeError](link), [Applicative](link), [Bifunctor](link), [Eq](link), [Foldable](link), [Functor](link), [MonadError](link), [Monad](link), [Monoid](link), [Semigroup](link), [SemigroupK](link), [Show](link), [Traverse](link)|
```

which renders as:

### Supported type classes

| Module | Type classes |
| --- | --- |
|__arrow.aql__|[Count](link), [From](link), [GroupBy](link), [OrderBy](link), [Select](link), [Sum](link), [Union](link)|
|__arrow.optics.typeclasses__|[Each](link)|
|__arrow.typeclasses__|[ApplicativeError](link), [Applicative](link), [Bifunctor](link), [Eq](link), [Foldable](link), [Functor](link), [MonadError](link), [Monad](link), [Monoid](link), [Semigroup](link), [SemigroupK](link), [Show](link), [Traverse](link)|

### CODE REVIEW

File: modules/ank/arrow-ank/src/main/kotlin/arrow/ank/interpreter.kt [hunk 1]
```diff
@@ -86,11 +86,12 @@ data class CompiledMarkdown(val origin: File, val snippets: ListK<Snippet>)
 data class Snippet(
   val fence: String,
   val lang: String,
-  val silent: Boolean,
   val startOffset: Int,
   val endOffset: Int,
   val code: String,
-  val result: Option<String> = None)
+  val result: Option<String> = None,
+  val isSilent: Boolean = fence.startsWith("```$lang$AnkSilentBlock"),
```
Comment (pakoito) [inline modules/ank/arrow-ank/src/main/kotlin/arrow/ank/interpreter.kt]: Can you make a `sealed class` out of this?

**Comment by pakoito** [comment]
What link does it produce at the moment?

**Comment by raulraja** [comment]
@pakoito Ank needs an overhaul where we move into `Effect` and out of `Free` and we will then rework all those data types as ADTs. //cc @Guardiola31337 

Regarding the links produced they are currently based on the data type or type class package and that is why I created earlier #1120

File: modules/docs/arrow-docs/docs/docs/aql/from/README.md [hunk 2]
```diff
@@ -45,6 +45,15 @@ Learn more about the `AQL` combinators
 - [_sum_](/docs/aql/sum/)
 - [_union_](/docs/aql/union/)
 
+### Supported Data types
+
+```kotlin:ank:replace
+import arrow.reflect.*
+import arrow.aql.*
+
+TypeClass(From::class).dtMarkdownList()
```
Comment (JorgeCastilloPrz) [inline modules/docs/arrow-docs/docs/docs/aql/from/README.md]: 🎉

File: modules/docs/arrow-docs/src/main/kotlin/ExtensionsHelper.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
+package arrow.reflect
+
+import kotlin.reflect.KClass
+
+private val lineSeparator: String = System.getProperty("line.separator")
+
+/**
+ * @return a list of [TypeClass] supported by this [DataType]
+ */
+fun DataType.tcMarkdownList(): String =
```
Comment (JorgeCastilloPrz) [inline modules/docs/arrow-docs/src/main/kotlin/ExtensionsHelper.kt]: It's incredible how simple it is 👌

**Pull request merged**